README.md
# Azure IoT Hub with TOPPERS/ASP3

RTOSの[TOPPERS/ASP3](https://www.toppers.jp/asp3-kernel.html)を使ったAzure IoT Hub へ接続する[Azure IoT Hub Device C SDK](https://github.com/Azure/azure-iot-sdk-c)を使ったサンプルです。
ターゲットデバイスは、[GR-SAKURA](http://gadget.renesas.com/ja/product/sakura.html)で、Ethernet上のHTTPかMQTTで通信します。 MQTTは、mbedTLSを使用した場合に接続できます。

[GR-PEACH](http://gadget.renesas.com/ja/product/peach.html)でも動作確認できています。ベースのソフトに[TOPPERS/ASP3版ECNL](http://dev.toppers.jp/trac_user/contrib/wiki/asp3_tinet_ecnl_rx)を使用します。注意点として、「get_time」という関数が「c-utility/adapters/agenttime_mbed.c」に定義されていますが、ベースのソフトにも定義されているので、Azure IoT SDK側の「get_time」関数を「azure_get_time」などにリネームして競合を避ける必要があります。

サンプルを動作させるには、[この手順](https://docs.microsoft.com/ja-jp/azure/iot-hub/quickstart-send-telemetry-c)や[この手順](https://github.com/ms-iotkithol-jp/IoTKitHoLV4)を参照にAzure側にデバイスを追加し、そのデバイスの接続文字列を取得し、ソースコード「app_iothub_client/src/client.c」の以下の個所を書き換える必要があります。

```c
static const char* connectionString = "[device connection string]";
```

## 開発環境

ビルドはGCC+Makefileで行います。RX向けGCCは[こちら](https://gcc-renesas.com/ja/rx-download-toolchains/)から入手できます。
IDEとして[e² studio V7.4.0](https://www.renesas.com/ja-jp/products/software-tools/tools/ide/e2studio.html)を使用しました。
Rubyの実行環境が必要なので、​[こちら](https://www.ruby-lang.org/ja/downloads/)からダウンロードして、コマンドプロントから実行できるよう、環境変数でパスを通してください。
e² studioのプロジェクトファイルも含まれているので、展開したフォルダをワークスペースとして開き、プロジェクトのインポートをこのフォルダに対して行うことで、取り込むことができます。
開発手順は[ここ](https://dev.toppers.jp/trac_user/contrib/wiki/azure_iot_hub)にあります。

## フォルダ構成

|パス|内容|ライセンス|
|:---|:---|:---|
|app_iothub_client|サンプルプログラム|TOPPERS License|
|asp3_dcre|RTOS [TOPPERS/ASP3](http://www.toppers.jp/asp3-kernel.html)|TOPPERS License|
|asp3_dcre/tinet|TCPプロトコルスタック [TINET 1.7](http://www.toppers.jp/tinet.html)|TOPPERS License|
|asp3_dcre/mbed|[​mbed](https://www.mbed.com/) HAL APIの利用と実装|Apache 2.0 License|
|asp3_dcre/syssvc/tlsf|メモリアロケータ [tlsf](https://github.com/mattconte/tlsf)|BSD License|
|azure_iothub|[Azure IoT Hub Device C SDK](https://github.com/Azure/azure-iot-sdk-c) (tag 2019-04-11)|MIT License|
|azure_iothub/deps/parson|jsonパーサー [parson](https://github.com/kgabis/parson)|MIT License|
|azure_iothub/c-utility/adapters|TOPPERS向けの実装を追加|TOPPERS License|
|curl-7.57.0|HTTPクライアント[cURL](https://curl.haxx.se/)|Curl License|
|musl-1.1.18|標準Cライブラリ[musl](http://www.musl-libc.org/)|MIT License|
|ntshell/fatfs|FATファイルシステム [FatFs](http://elm-chan.org/fsw/ff/00index_e.html)|FatFs License|
|ntshell/ntshell|仮想ターミナル [NT-Shell](https://www.cubeatsystems.com/ntshell/)|MIT License etc.|
|ntshell/src|Socket API、標準入出力、ファイルディスクリプタ|TOPPERS License|
|wolfssl-3.15.7|SSL/TLS ライブラリ [wolfSSL](https://www.wolfssl.com/) 現在未使用|GPL/商用|
|mbedtls-2.16.1|SSL/TLS ライブラリ [mdebTLS](https://tls.mbed.org/)|Apache 2.0 License|
|zlib-1.2.11|圧縮・伸張ライブラリ [zlib](https://www.zlib.net/)|zlib License|

ライセンスは参考のため載せておきました。正確なライセンスは各ソフトウェアのサイトを確認してください。

|確認|プロトコル|ソフトスタック|
|-|-|-|
|:heavy_check_mark:|HTTP|iothubtransporthttp.c→httpapi_compact.c→tlsio_wolfssl.c→socketio_berkeley.c|
|:heavy_multiplication_x:|HTTP|iothubtransporthttp.c→httpapi_compact.c→tlsio_mbedtls.c→socketio_berkeley.c|
|:heavy_check_mark:|HTTP|iothubtransporthttp.c→httpapi_curl.c→Curl＋WolfSSL|
|:heavy_multiplication_x:|HTTP|iothubtransporthttp.c→httpapi_curl.c→Curl＋mbedSSL|
|:heavy_multiplication_x:|MQTT|iothubtransportmqtt.c→wsio.c→umqtt→tlsio_wolfssl.c→socketio_berkeley.c|
|:heavy_check_mark:|MQTT|iothubtransportmqtt.c→wsio.c→umqtt→tlsio_mbedtls.c→socketio_berkeley.c|
|:heavy_multiplication_x:|MQTT over WebSocket|iothubtransportmqtt_websockets.c→wsio.c→umqtt→tlsio_wolfssl.c→socketio_berkeley.c|
|:heavy_check_mark:|MQTT over WebSocket|iothubtransportmqtt_websockets.c→wsio.c→umqtt→tlsio_mbedtls.c→socketio_berkeley.c|
|:heavy_minus_sign:|AMQP||
|:heavy_minus_sign:|AMQP over WebSocket||

デフォルトでは、「MQTT/mbedTLS」が有効になっています。

### wolfSSLに変更するには

app_iothub_client\Debug\Makefile(60)
```
$(SRCDIR)/../mbedtls-2.16.1/Debug/libmbedtls.a \
```
を
```
$(SRCDIR)/../wolfssl-3.15.7/Debug/libwolfssl.a \
```
に変更。

「azure_iothub」プロジェクトのコンパイルオプションに、「USE_WOLFSSL=1」を指定

azure_iothub/c-utility/adapters/platform_toppers.c(55)
```c
return tlsio_mbedtls_get_interface_description();
```
を
```c
return tlsio_wolfssl_get_interface_description();
```
に変更。

curl-7.57.0/lib/curl_config.h(930)
```c
/* #undef USE_CYASSL */
```
を
```c
#define USE_CYASSL 1
```
に変更。

curl-7.57.0/lib/curl_config.h(954)
```c
#define USE_MBEDTLS 1
```
を
```c
/* #undef USE_MBEDTLS */
```
に変更。

## TOPPERS License

TOPPERS Licenseは、BSDライクなライセンスです。詳しくは、下記のサイトを見てください。

https://www.toppers.jp/license.html


curl-7.57.0/RELEASE-NOTES
Curl and libcurl 7.57.0

 Public curl releases:         171
 Command line options:         211
 curl_easy_setopt() options:   249
 Public functions in libcurl:  74
 Contributors:                 1649

This release includes the following changes:

 o auth: add support for RFC7616 - HTTP Digest access authentication [12]
 o share: add support for sharing the connection cache [31]
 o HTTP: implement Brotli content encoding [28]

This release includes the following bugfixes:

 o CVE-2017-8816: NTLM buffer overflow via integer overflow [47]
 o CVE-2017-8817: FTP wildcard out of bounds read [48]
 o CVE-2017-8818: SSL out of buffer access [49]
 o curl_mime_filedata.3: fix typos [1]
 o libtest: Add required test libraries for lib1552 and lib1553 [2]
 o fix time diffs for systems using unsigned time_t [3]
 o ftplistparser: memory leak fix: free temporary memory always [4]
 o multi: allow table handle sizes to be overridden [5]
 o wildcards: don't use with non-supported protocols [6]
 o curl_fnmatch: return error on illegal wildcard pattern [7]
 o transfer: Fix chunked-encoding upload too early exit [8]
 o curl_setup: Improve detection of CURL_WINDOWS_APP [9]
 o resolvers: only include anything if needed [10]
 o setopt: fix CURLOPT_SSH_AUTH_TYPES option read
 o appveyor: add a win32 build
 o Curl_timeleft: change return type to timediff_t [11]
 o cmake: Export libcurl and curl targets to use by other cmake projects [13]
 o curl: in -F option arg, comma is a delimiter for files only [14]
 o curl: improved ";type=" handling in -F option arguments
 o timeval: use mach_absolute_time() on MacOS [15]
 o curlx: the timeval functions are no longer provided as curlx_* [16]
 o mkhelp.pl: do not generate comment with current date [17]
 o memdebug: use send/recv signature for curl_dosend/curl_dorecv [18]
 o cookie: avoid NULL dereference [19]
 o url: fix CURLOPT_POSTFIELDSIZE arg value check to allow -1 [20]
 o include: remove conncache.h inclusion from where its not needed
 o CURLOPT_MAXREDIRS: allow -1 as a value [21]
 o tests: Fixed torture tests on tests 556 and 650
 o http2: Fixed OOM handling in upgrade request
 o url: fix CURLOPT_DNS_CACHE_TIMEOUT arg value check to allow -1
 o CURLOPT_INFILESIZE: accept -1 [22]
 o curl: pass through [] in URLs instead of calling globbing error [23]
 o curl: speed up handling of many URLs [24]
 o ntlm: avoid malloc(0) for zero length passwords [25]
 o url: remove faulty arg value check from CURLOPT_SSH_AUTH_TYPES [26]
 o HTTP: support multiple Content-Encodings [27]
 o travis: add a job with brotli enabled
 o url: remove unncessary NULL-check
 o fnmatch: remove dead code
 o connect: store IPv6 connection status after valid connection [29]
 o imap: deal with commands case insensitively [30]
 o --interface: add support for Linux VRF [32]
 o content_encoding: fix inflate_stream for no bytes available [33]
 o cmake: Correctly include curl.rc in Windows builds [34]
 o cmake: Add missing setmode check [35]
 o connect.c: remove executable bit on file [36]
 o SMB: fix uninitialized local variable
 o zlib/brotli: only include header files in modules needing them [37]
 o URL: return error on malformed URLs with junk after IPv6 bracket [38]
 o openssl: fix too broad use of HAVE_OPAQUE_EVP_PKEY [39]
 o macOS: Fix missing connectx function with Xcode version older than 9.0 [40]
 o --resolve: allow IP address within [] brackets [41]
 o examples/curlx: Fix code style [42]
 o ntlm: remove unnecessary NULL-check to please scan-build [43]
 o Curl_llist_remove: fix potential NULL pointer deref [43]
 o mime: fix "Value stored to 'sz' is never read" scan-build error [43]
 o openssl: fix "Value stored to 'rc' is never read" scan-build error [43]
 o http2: fix "Value stored to 'hdbuf' is never read" scan-build error [43]
 o http2: fix "Value stored to 'end' is never read" scan-build error [43]
 o Curl_open: fix OOM return error correctly [43]
 o url: reject ASCII control characters and space in host names [44]
 o examples/rtsp: clear RANGE again after use [45]
 o connect: improve the bind error message [46]
 o make: fix "make distclean" [50]
 o connect: add support for new TCP Fast Open API on Linux [51]
 o metalink: fix memory-leak and NULL pointer dereference [52]
 o URL: update "file:" URL handling [53]
 o ssh: remove check for a NULL pointer [54]
 o global_init: ignore CURL_GLOBAL_SSL's absense [55]

This release includes the following known bugs:

 o see docs/KNOWN_BUGS (https://curl.haxx.se/docs/knownbugs.html)

This release would not have looked like this without help, code, reports and
advice from friends like these:

  Alessandro Ghedini, Alex Malinovich, Alex Nichols, Alfonso Martone,
  Andrew Lambert, arainchik on github, Brian Carpenter, cbartl on github,
  Dan Fandrich, Daniel Bankhead, Daniel Stenberg, Dirk Feytons,
  Dmitri Tikhonov, Evgeny Grin, Gisle Vanem, hsiao yi, Jakub Zakrzewski,
  John Starks, Juro Bystricky, Kamil Dudka, Luca Boccassi, Marcel Raad,
  Martin Storsjö, Matthew Kerwin, Max Dymond, Michael Felt, Michael Kaufmann,
  moohoorama on github, omau on github, Orgad Shaneh, Patrick Monnerat,
  Paul Howarth, Pavel Gushchin, Pavol Markovic, Per Lundberg, Peter Piekarski,
  Petr Voytsik, Ray Satiro, Rob Cotrone, Viktor Szakáts, youngchopin on github,
  (41 contributors)

        Thanks! (and sorry if I forgot to mention someone)

References to bug reports and discussions on issues:

 [1] = https://curl.haxx.se/bug/?i=2008
 [2] = https://curl.haxx.se/bug/?i=2006
 [3] = https://curl.haxx.se/bug/?i=2004
 [4] = https://curl.haxx.se/bug/?i=2013
 [5] = https://curl.haxx.se/bug/?i=1982
 [6] = https://curl.haxx.se/bug/?i=2016
 [7] = https://curl.haxx.se/bug/?i=2015
 [8] = https://curl.haxx.se/bug/?i=2001
 [9] = https://curl.haxx.se/bug/?i=2025
 [10] = https://curl.haxx.se/bug/?i=2023
 [11] = https://curl.haxx.se/bug/?i=2021

asp3_dcre/configure.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-
#
#  TOPPERS Software
#      Toyohashi Open Platform for Embedded Real-Time Systems
# 
#  Copyright (C) 2001-2003 by Embedded and Real-Time Systems Laboratory
#                              Toyohashi Univ. of Technology, JAPAN
#  Copyright (C) 2006-2019 by Embedded and Real-Time Systems Laboratory
#              Graduate School of Information Science, Nagoya Univ., JAPAN
# 
#  上記著作権者は，以下の(1)～(4)の条件を満たす場合に限り，本ソフトウェ
#  ア（本ソフトウェアを改変したものを含む．以下同じ）を使用・複製・改
#  変・再配布（以下，利用と呼ぶ）することを無償で許諾する．
#  (1) 本ソフトウェアをソースコードの形で利用する場合には，上記の著作
#      権表示，この利用条件および下記の無保証規定が，そのままの形でソー
#      スコード中に含まれていること．
#  (2) 本ソフトウェアを，ライブラリ形式など，他のソフトウェア開発に使
#      用できる形で再配布する場合には，再配布に伴うドキュメント（利用
#      者マニュアルなど）に，上記の著作権表示，この利用条件および下記
#      の無保証規定を掲載すること．
#  (3) 本ソフトウェアを，機器に組み込むなど，他のソフトウェア開発に使
#      用できない形で再配布する場合には，次のいずれかの条件を満たすこ
#      と．
#    (a) 再配布に伴うドキュメント（利用者マニュアルなど）に，上記の著
#        作権表示，この利用条件および下記の無保証規定を掲載すること．
#    (b) 再配布の形態を，別に定める方法によって，TOPPERSプロジェクトに
#        報告すること．
#  (4) 本ソフトウェアの利用により直接的または間接的に生じるいかなる損
#      害からも，上記著作権者およびTOPPERSプロジェクトを免責すること．
#      また，本ソフトウェアのユーザまたはエンドユーザからのいかなる理
#      由に基づく請求からも，上記著作権者およびTOPPERSプロジェクトを
#      免責すること．
# 
#  本ソフトウェアは，無保証で提供されているものである．上記著作権者お
#  よびTOPPERSプロジェクトは，本ソフトウェアに関して，特定の使用目的
#  に対する適合性も含めて，いかなる保証も行わない．また，本ソフトウェ
#  アの利用により直接的または間接的に生じたいかなる損害に関しても，そ
#  の責任を負わない．
# 
#  $Id: configure.rb 1270 2019-10-03 14:04:50Z ertl-hiro $
# 

Encoding.default_external = 'utf-8'
require "optparse"
require "fileutils"
require "shell"

#  オプションの定義
#
#  -T <target>			ターゲット名（必須）
#  -a <appldirs>		アプリケーションのディレクトリ名（複数指定可．デ
#						フォルトはsampleディレクトリ）
#  -A <applname>		アプリケーションプログラム名（デフォルトはsample1）
#  -t					メインのオブジェクトファイルをリンク対象に含めない
#  -c <cfgfile>			システムコンフィギュレーションファイル（.cfgファイ
#						ル名）名
#  -C <cdlflle>			コンポーネント記述ファイル（.cdlファイル）名
#  -U <applobjs>		アプリケーションの追加のオブジェクトファイル
#						（.oファイル名で指定．複数指定可）
#  -S <syssvcobjs>		システムサービスのオブジェクトファイル
#						（.oファイル名で指定．複数指定可）
#  -B <bannerobj>		バナー表示のオブジェクトファイル（.oファイル名で指定）
#  -L <kernel_lib>		カーネルライブラリ（libkernel.a）のディレクトリ名
#						（省略した場合，カーネルライブラリもmakeする）
#  -f					カーネルを関数単位でコンパイルするかどうかの指定
#  -D <srcdir>			カーネルソースの置かれているディレクトリ
#  -l <srclang>			プログラミング言語（現時点ではcとc++のみサポート）
#  -m <tempmakefile>	Makefileのテンプレートのファイル名の指定（デフォル
#						トはsampleディレクトリのMakefile）
#  -d <objdir>			中間オブジェクトファイルと依存関係ファイルを置く
#						ディレクトリ名（デフォルトはobjs）
#  -w					TECSを使用しない
#  -W <tecsdir>			TECS関係ファイルのディレクトリ名（デフォルトはソー
#						スファイルのディレクトリの下のtecsgen）
#  -r					トレースログ記録のサンプルコードを使用するかどうか
#						の指定
#  -V <devtooldir>		開発ツール（コンパイラ等）の置かれているディレクトリ
#  -R <ruby>			rubyのパス名（明示的に指定する場合）
#  -g <cfg>				コンフィギュレータ（cfg）のパス名
#  -G <tecsgen>			TECSジェネレータ（tecsgen）のパス名
#  -o <options>			コンパイルオプション（COPTSに追加）
#  -O <options>			シンボル定義オプション（CDEFSに追加）
#  -k <options>			リンカオプション（LDFLAGSに追加）
#  -b <options>			リンカオプション（LIBSに追加）
#  -e <tinetdir>		TINET のソースの置かれているディレクトリ
#  -i <net_if>			ネットワークインタフェース（TINETが有効の場合は必須）
#						ether、ppp、loop の何れかを指定する。
#  -v <net_dev>			イーサネット・ディバイスドライバ
#  						（ネットワークインタフェースに ether を指定した場合は必須）
#  -n <net_proto>		ネットワーク層プロトコル（TINETが有効の場合は必須）
#						inet6、inet4、inet6/inet4、inet6m、inet6m/inet4 の何れかを指定する。
#  -s <trans_proto>		トランスポート層プロトコル（TINETが有効の場合は必須）
#						tcp、udp、tcp/udp の何れかを指定する。

#  使用例(1)
#
#  % ../configure.rb -T ct11mpcore_gcc -O "-DTOPPERS_USE_QEMU" \
#					-A perf1 -a ../test -S "test_svc.o histogram.o"
#
#  使用例(2)
#
#  % ../configure.rb -T macosx_gcc -L .
#	アプリケーションプログラムは，sample1になる．
#
#  使用例(3)
#
#  % ../configure.rb -T ct11mpcore_gcc -O "-DTOPPERS_USE_QEMU" -A tSample2 -t
#	アプリケーションプログラムは，TECS版のサンプルプログラムになる．
#
#  使用例(4)
#
#  % ../configure.rb -T ct11mpcore_gcc PRC_NUM=4
#	PRC_NUMを4に定義する．
#
#  使用例(5)
#
#  % ../configure.rb -T gr_sakura_gcc -A sample1n \
#					-e tinet -i ether -v if_rx62n -n inet4 -s tcp -U sample1.o
#	TINETを有効にする．

#
#  変数の初期化
#
$target = nil
$appldirs = []
$applname = nil
$option_t = false
$cfgfile = nil
$cdlfile = nil
$applobjs = []
$syssvcobjs = []
$bannerobj = nil
$kernel_lib = ""
$kernel_funcobjs = ""
$srcdir = nil
$srclang = "c"
$tempmakefile = nil
$objdir = "objs"
$omit_tecs = ""
$tecsdir = nil
$enable_trace = ""
$devtooldir = ""
$ruby = "ruby"
$cfg = nil
$tecsgen = nil
$copts = []
$cdefs = []
$ldflags = []
$libs = []

#
#  オプションの処理
#
OptionParser.new(nil, 22) do |opt|
  opt.on("-T target",		"taget name (mandatory)") do |val|
    $target = val
  end
  opt.on("-a appldirs",		"application directories") do |val|
    $appldirs += val.split(/\s+/)
  end
  opt.on("-A applname",		"application program name") do |val|
    $applname = val
  end
  opt.on("-t",				"TECS is used for application development") do |val|
    $option_t = true
  end
  opt.on("-c cfgfile",		"system configuration file") do |val|
    $cfgfile = val
  end
  opt.on("-C cdlfile",		"component description file") do |val|
    $cdlfile = val
  end
  opt.on("-U applobjs",		"additional application object files") do |val|
    $applobjs += val.split(/\s+/)
  end
  opt.on("-S syssvcobjs",	"system service object files") do |val|
    $syssvcobjs += val.split(/\s+/)
  end
  opt.on("-B bannerobj",	"banner display object file") do |val|
    $bannerobj = val
  end
  opt.on("-L kernel_lib",	"directory of built kernel library") do |val|
    $kernel_lib = val
  end
  opt.on("-f", "each function is complied separately in kernel") do |val|
    $kernel_funcobjs = "true"
  end
  opt.on("-D srcdir",		"path of source code directory") do |val|
    $srcdir = val
  end
  opt.on("-l srclang",		"programming language (C or C++)") do |val|
    $srclang = val
  end
  opt.on("-m tempmakefile", "template file of Makefile") do |val|
    $tempmakefile = val
  end
  opt.on("-d objdir",		"relocatable object file directory") do |val|
    $objdir = val
  end
  opt.on("-w",				"TECS is not used at all") do |val|
    $omit_tecs = "true"
  end
  opt.on("-W tecsdir",		"path of TECS file directory") do |val|
    $tecsdir = val
  end
  opt.on("-r",				"use the sample code for trace log") do |val|
    $enable_trace = "true"
  end
  opt.on("-V devtooldir",	"development tools directory") do |val|
    $devtooldir = val
  end
  opt.on("-R ruby",			"path of ruby command") do |val|
    $ruby = val
  end
  opt.on("-g cfg",			"path of configurator") do |val|
    $cfg = val
  end
  opt.on("-G tecsgen",		"path of TECS generator") do |val|
    $tecsgen = val
  end
  opt.on("-o options",		"compiler options") do |val|
    $copts += val.split(/\s+/)
  end
  opt.on("-O options",		"symbol definition options") do |val|
    $cdefs += val.split(/\s+/)
  end
  opt.on("-k options",		"linker options") do |val|
    $ldflags += val.split(/\s+/)
  end
  opt.on("-b options",		"linker options for linking libraries") do |val|
    $libs += val.split(/\s+/)
  end
  opt.on("-e options",		"path of TINET directory") do |val|
    $tinetdir = val
  end
  opt.on("-i options",		"network interface (ether or ppp or loop)") do |val|
    $net_if = val
  end
  opt.on("-v options",		"ethernet device driver") do |val|
    $net_dev = val
  end
  opt.on("-n options",		"network layer protocol (inet6 or inet4 or inet6/inet4 or inet6m or inet6m/inet4)") do |val|
    $net_proto = val
  end
  opt.on("-s options",		"transeport layer protocol (tcp or udp or tcp/udp)") do |val|
    $trans_proto = val
  end
  opt.parse!(ARGV)
end

#
#  オブジェクトファイル名の拡張子を返す
#
def GetObjectExtension
  if /cygwin/ =~ RUBY_PLATFORM
    return("exe")
  else
    return("")
  end
end

#
#  変数のデフォルト値（文字列変数のデフォルト値は初期化で与える）
#
if $appldirs.empty?
  $appldirs.push("\$(SRCDIR)/sample")
end
$applname ||= "sample1"
$cfgfile ||= $applname + ".cfg"
$cdlfile ||= $applname + ".cdl"
$applobjs.unshift($applname + ".o") if !$option_t
$bannerobj ||= ($omit_tecs == "") ? "tBannerMain.o" : "banner.o"
if $srcdir.nil?
  # ソースディレクトリ名を取り出す
  if /(.*)\/configure/ =~ $0
    $srcdir = $1
  else
    $srcdir = Shell.new.cwd
  end
end
if /^\// =~ $srcdir
  $srcabsdir = $srcdir
else
  $srcabsdir = Shell.new.cwd + "/" + $srcdir
end
$tempmakefile ||= $srcdir + "/sample/Makefile"
$tecsdir ||= "\$(SRCDIR)/tecsgen"
$cfg ||= $ruby + " \$(SRCDIR)/cfg/cfg.rb"
$tecsgen ||= $ruby + " \$(TECSDIR)/tecsgen.rb"

#
#  -Tオプションとターゲット依存部ディレクトリの確認
#
if $target.nil?
  puts("configure.rb: -T option is mandatory")
elsif !File.directory?($srcdir + "/target/" + $target)
  puts("configure.rb: #{$srcdir}/target/#{$target} not exist.")
  $target = nil
end

if $target.nil?
  puts("Installed targets are:")
  Dir.glob($srcdir + "/target/[a-zA-Z0-9]*").each do |target|
    target.sub!($srcdir + "/target/", "")
    puts("\t" + target)
  end
  abort
end

#
#  変数テーブルの作成
#
$vartable = Hash.new("")
$vartable["TARGET"] = $target
$vartable["APPLDIRS"] = $appldirs.join(" ")
$vartable["APPLNAME"] = $applname
$vartable["CFGFILE"] = $cfgfile
$vartable["CDLFILE"] = $cdlfile
$vartable["APPLOBJS"] = $applobjs.join(" ")
$vartable["SYSSVCOBJS"] = $syssvcobjs.join(" ")
$vartable["BANNEROBJ"] = $bannerobj
$vartable["KERNEL_LIB"] = $kernel_lib
$vartable["KERNEL_FUNCOBJS"] = $kernel_funcobjs
$vartable["SRCDIR"] = $srcdir
$vartable["SRCABSDIR"] = $srcabsdir
$vartable["SRCLANG"] = $srclang
$vartable["OBJDIR"] = $objdir
$vartable["OMIT_TECS"] = $omit_tecs
$vartable["TECSDIR"] = $tecsdir
$vartable["ENABLE_TRACE"] = $enable_trace
$vartable["DEVTOOLDIR"] = $devtooldir
$vartable["RUBY"] = $ruby
$vartable["CFG"] = $cfg
$vartable["TECSGEN"] = $tecsgen
$vartable["COPTS"] = $copts.join(" ")
$vartable["CDEFS"] = $cdefs.join(" ")
$vartable["LDFLAGS"] = $ldflags.join(" ")
$vartable["LIBS"] = $libs.join(" ")
$vartable["OBJEXT"] = GetObjectExtension()
$vartable["TINETDIR"] = $tinetdir
$vartable["NET_IF"] = $net_if
$vartable["NET_DEV"] = $net_dev
$vartable["SUPPORT_INET6"] = ($net_proto == "inet6") || ($net_proto == "inet6/inet4") || ($net_proto == "inet6m") || ($net_proto == "inet6m/inet4")
$vartable["SUPPORT_INET4"] = ($net_proto == "inet4") || ($net_proto == "inet6/inet4") || ($net_proto == "inet6m/inet4")
$vartable["SUPPORT_TCP"] = ($trans_proto == "tcp") || ($net_proto == "tcp/udp")
$vartable["SUPPORT_UDP"] = ($trans_proto == "udp") || ($net_proto == "tcp/udp")
$vartable["API_CFG_IP4MAPPED_ADDR"] = ($net_proto == "inet6m") || ($net_proto == "inet6m/inet4")

ARGV.each do |arg|
  if /^([A-Za-z0-9_]+)\s*\=\s*(.*)$/ =~ arg
    $vartable[$1] = $2
  else
    $vartable[arg] = true
  end
end

#
#  ファイルを変換する
#
def convert(inFileName, outFileName)
  puts("Generating #{outFileName} from #{inFileName}.\n")
  if (File.file?(outFileName))
    puts("#{outFileName} exists.  Save as #{outFileName}.bak.\n")
    FileUtils.move(outFileName, outFileName + ".bak")
  end

  begin
    inFile = File.open(inFileName)
    outFile = File.open(outFileName, "w")
  rescue Errno::ENOENT, Errno::EACCES => ex
    abort(ex.message)
  end

  inFile.each do |line|
    line.chomp!
    while line.sub!(/\@\(([A-Za-z0-9_]+)\)/) {|var| $vartable[$1]} do end
    outFile.print(line,"\n")
  end

  inFile.close
  outFile.close
end

#
#  Makefileの生成
#
convert($tempmakefile, "Makefile")

#
#  依存関係ファイルのディレクトリの作成
#
if !File.directory?($objdir)
  Dir.mkdir($objdir)
end
